modules/packages/ParallelIO.chpl

```diff
@@ -0,0 +1,705 @@
+    to open the file
+  */
+  proc readParallelLocal(filePath: string, type t, in delim: ?dt = b"\n",
+                         header = headerPolicy.noHeader, nTasks: int = here.maxTaskPar,
+                         type deserializerType = defaultDeserializer
+  ): [] t throws
+    where dt == bytes || dt == string
+  {
+    var results: [0..nTasks] list(t);
+    const f = open(filePath, ioMode.r),
+          fileBounds = 0..<f.size,
+          byteOffsets = findDelimChunksChecked(f, delim, nTasks, t, fileBounds,
+                                               header, deserializerType);
+
+    coforall tid in 0..<nTasks with (ref results) {
+      const taskBounds = byteOffsets[tid]..byteOffsets[tid+1];
+      var des: deserializerType,
+          r = f.reader(locking=false, region=taskBounds, deserializer=des),
+          s = new t();
+
+      // read all the 't' values in the chunk into a list
+      while r.read(s) do
+        results[tid].pushBack(s);
+    }
+
+    // gather results from each task into an array
+    const nPerTask = [r in results] r.size,
+          itemOffsets = (+ scan nPerTask) - nPerTask;
+
+    var result: [0..<itemOffsets.last] t;
+    coforall tid in 0..<nTasks do
+      result[itemOffsets[tid]..<itemOffsets[tid+1]] = results[tid].toArray();
+
+    return result;
+  }
+
+  /*
+    Get an array of ``n+1`` byte offsets that divide the file ``f`` into ``n``
+    roughly equally sized chunks, where each byte offset lines up with a
+    delimiter.
+
+    :arg f: the file to search
+    :arg delim: the delimiter to use to separate the file into chunks
+    :arg n: the number of chunks to find
+    :arg bounds: a range of byte offsets to break into chunks
+    :arg header: a header policy to use when searching for the first byte offset
+
+    :returns: a length ``n+1`` array of byte offsets (the last offset is
+              ``bounds.high``)
+
+    :throws: :class:`OffsetNotFoundError` if a valid byte offset cannot be found
+              in any of the chunks
+  */
+  proc findDelimChunks(const ref f: file, in delim: ?dt, n: int, bounds: range,
+                       header = headerPolicy.noHeader
+  ): [] int throws
+    where dt == bytes || dt == string
+  {
+    const (findStart, startOffset) = header.apply(f, bounds),
+          chunkIndices = if findStart then 0..<n else 1..<n;
+
+    const nDataBytes = bounds.high - startOffset,
+          approxBytesPerChunk = nDataBytes / n;
+
+    var chunkOffsets: [0..n] int;
+    chunkOffsets[0] = startOffset;
+    chunkOffsets[n] = bounds.high;
+
+    coforall i in chunkIndices with (ref chunkOffsets) {
+      const estOffset = chunkOffsets[0] + i * approxBytesPerChunk,
+            r = f.reader(locking=false, region=estOffset..);
+
+      try {
+        r.advanceThrough(delim); // advance past the next delimiter
+      } catch {
+        // there wasn't an offset in this chunk
+        throw new OffsetNotFoundError(
```

Comment:
IMO this wouldn't need to throw if the last element doesn't have the delimiter. E.g. a file with a bunch of lines and then no newline in the last line.

Reply:
The last offset is set to the last offset of the provided range:
```chapel
chunkOffsets[n] = bounds.high;
```
which isn't checked in the loop.

So there is no requirement that there is a delimiter at the very end of the file, only that the last chunk has a delimiter. In other words, if `n=10` and `bounds=1..100`, it will throw if there isn't a delimiter anywhere in `90..100`.

Reply:
That's interesting. I wonder if it'd be reasonable to catch this error and then fall back on serial code or something. I think this error could come up with sufficiently strange (but valid) input, so I find it concerning.

Reply:
> I think this error could come up with sufficiently strange (but valid) input, so I find it concerning.

Using too many tasks to read a small file would be the main case I can think of where you'd hit this error (other than a corrupted file or something), because you could end up with a `t` value that spans multiple chunks.

Are you seeing another case where reducing the number of tasks wouldn't resolve the issue?

Also, I agree that falling back to a serial read would be much nicer than throwing in general but I haven't gotten to implementing that yet.